### PR TITLE
Speed up migration by making from.Get() and to.Put() concurrent

### PR DESCRIPTION
**Pull request by philipsu522**
This PR addresses the issue in https://github.com/hashicorp/vault/issues/8074 to speed up the migration.

This is achieved by starting goroutines for from.Get() and from.Put(), and limiting the number of concurrent migrations via a semaphore.

I ran unit tests and verified that they passed, as well as acceptance tests.
Unit tests:
```
=== RUN   TestMigration
--- PASS: TestMigration (1.37s)
=== RUN   TestMigration/Default
    --- PASS: TestMigration/Default (0.75s)
=== RUN   TestMigration/Concurrent_migration
    --- PASS: TestMigration/Concurrent_migration (0.48s)
=== RUN   TestMigration/Start_option
    --- PASS: TestMigration/Start_option (0.13s)
=== RUN   TestMigration/Config_parsing
    --- PASS: TestMigration/Config_parsing (0.00s)
=== RUN   TestMigration/DFS_Scan
    --- PASS: TestMigration/DFS_Scan (0.00s)
=== RUN   TestMigration/Concurrent_DFS_Scan_limits_number_of_goroutines
    --- PASS: TestMigration/Concurrent_DFS_Scan_limits_number_of_goroutines (0.00s)
PASS
```
Acceptance tests:
```
make testacc TEST=./builtin/logical/consul
...
PASS
ok  	github.com/hashicorp/vault/builtin/logical/consul	30.518s
```

Deployed this version and migrated with concurrency.
With concurrency=1 it took about 41s to migrate all the files under "auth/<sha>/cert":
```
/ #  vault operator migrate -config migrate.hcl -concurrency 1
2020-01-07T01:02:46.131Z [WARN]  appending trailing forward slash to path
2020-01-07T01:05:31.640Z [INFO]  copied key: path=auth/63fe6b1c-9fd8-00cf-727e-0df237ec2e1f/cert/central-kube-bootstrap-aws-us-east-2
...
2020-01-07T01:03:12.143Z [INFO]  copied key: path=auth/63fe6b1c-9fd8-00cf-727e-0df237ec2e1f/cert/serverless
```
With concurrency=24  it took about .03 seconds to migrate all certs
```
/ # vault operator migrate -config migrate.hcl -concurrency 24
2020-01-07T01:05:29.233Z [WARN]  appending trailing forward slash to path
2020-01-07T01:05:31.640Z [INFO]  copied key: path=auth/63fe6b1c-9fd8-00cf-727e-0df237ec2e1f/cert/central-kube-bootstrap-aws-us-east-2
...
2020-01-07T01:05:34.894Z [INFO]  copied key: path=auth/63fe6b1c-9fd8-00cf-727e-0df237ec2e1f/cert/serverless
```

One thing to consider is that there is no longer strict ordering with the migrations.

### CODE REVIEW

File: command/operator_migrate.go [hunk 1]
```diff
@@ -199,27 +200,33 @@ func (c *OperatorMigrateCommand) migrate(config *migratorConfig) error {
 
 // migrateAll copies all keys in lexicographic order.
 func (c *OperatorMigrateCommand) migrateAll(ctx context.Context, from physical.Backend, to physical.Backend) error {
-	return dfsScan(ctx, from, func(ctx context.Context, path string) error {
-		if path < c.flagStart || path == storageMigrationLock || path == vault.CoreLockPath {
-			return nil
-		}
+	var wg = sync.WaitGroup{}
+	errorCh := make(chan error)
+	go func() {
```
Comment (kalafut) [inline command/operator_migrate.go]: Why use an error channel and invoke the call to `dfsScan` in a goroutine?

File: command/operator_migrate.go [hunk 2]
```diff
@@ -359,11 +366,17 @@ func dfsScan(ctx context.Context, source physical.Backend, cb func(ctx context.C
...
 		} else {
-			err := cb(ctx, key)
+			errorCh := make(chan error)
+			wg.Add(1)
+			go func() {
+				errorCh <- cb(ctx, key)
+				wg.Done()
+			}()
+
+			err := <- errorCh
```
Comment (kalafut) [inline command/operator_migrate.go]: This still keeps everything serialized, blocking until the callback returns. I don't think concurrent invocations of `cb` will be running as result

File: command/operator_migrate.go [hunk 3]
```diff
@@ -339,7 +346,7 @@ func parseStorage(result *migratorConfig, list *ast.ObjectList, name string) err
 
 // dfsScan will invoke cb with every key from source.
 // Keys will be traversed in lexicographic, depth-first order.
-func dfsScan(ctx context.Context, source physical.Backend, cb func(ctx context.Context, path string) error) error {
+func dfsScan(ctx context.Context, source physical.Backend, cb func(ctx context.Context, path string) error, wg sync.WaitGroup) error {
```
Comment (kalafut) [inline command/operator_migrate.go]: `wg` needs to be a `*sync.WaitGroup`, otherwise each call will get a copy and the top level `Wait` won't work.

**Comment by kalafut** [comment]
Hi. Parallelization is definitely something that has been discussed so thanks for the PR. I've noted some implementation issues, most basic of which that I don't believe the copy operations are running concurrently as expected. 

More generally, I think we need to review control the max concurrency, and whether that should be a user option. As is now you could end up trying to start thousands of goroutines and in turn simultaneous network connections. Some sort of batching or semaphore setup seems necessary, and this is something I'd like to discuss with the team when more people are back after the holidays.

**Comment by philipsu522** [comment]
> Hi. Parallelization is definitely something that has been discussed so thanks for the PR. I've noted some implementation issues, most basic of which that I don't believe the copy operations are running concurrently as expected.
> 
> More generally, I think we need to review how to control the max concurrency, and whether that should be a user option. As is now you could end up trying to start thousands of goroutines and in turn simultaneous network connections. Some sort of batching or semaphore setup seems necessary, and this is something I'd like to discuss with the team when more people are back after the holidays.

Thanks @kalafut, I've never written in Go before so the feedback was incredibly informative. While you discuss with the team exactly how to limit concurrency, I'll look at addressing your feedback for now

File: command/operator_migrate.go [hunk 4]
```diff
@@ -339,39 +369,61 @@ func parseStorage(result *migratorConfig, list *ast.ObjectList, name string) err
...
+	errorCh := make(chan error)
 	dfs := []string{""}
+	// This lock is used to coordinate modifications to dfs
+	dfsMutex := sync.Mutex{}
+	var wg = sync.WaitGroup{}
 
 	for l := len(dfs); l > 0; l = len(dfs) {
+		dfsMutex.Lock()
 		key := dfs[len(dfs)-1]
+		dfs = dfs[:len(dfs) - 1]
```
Comment (philipsu522) [inline command/operator_migrate.go]: moved truncation here so that we can immediately unlock the dfs mutex

**Comment by kalafut** [comment]
@philipsu522 Thanks for the updates. I'll be reviewing them soon.

**Comment by philipsu522** [comment]
@kalafut was wondering when you get some extra time if you could take a quick look. We're going to use a custom docker image built off of this PR to unblock ourselves, but would be great to have this feature available for others. Thanks!

**Comment by kalafut** [comment]
Hi @philipsu522. Thanks for your patience on this and glad you're unblocked with the custom build. There is definitely interest in getting it in, though we're continuing to review the migration process, in particular with the new (beta) integrated storage.

**Comment by wmyre** [comment]
Can we please get this added to a release? Migrations are painfully slow and I cannot find supporting documentation which describes what the 250,000 certificates in path=logical/{GUID}/certs/ are for.

**Pull request closed without merging**